Transfers: Add FailureRate SourceRankingStrategy

JChenster opened the pull request:
We add a SourceRankingStrategy that makes source selection dynamic with respect to failure rate. We look up files done/failed in the past hour in the TransferStats schema. We assign RSEs with lower failure rates higher priority.

<!-- Please read https://rucio.cern.ch/documentation/contributing before submitting a pull request -->

rcarpa commented on lib/rucio/core/transfer.py:
```diff
@@ -1071,6 +1071,69 @@ def apply(self, ctx: RequestRankingContext, source: RequestSource) -> "Optional[
+        return FailureRate._RankingContext(self, rws, failure_rate_for_rws)
+
+    def apply(self, ctx: RequestRankingContext, source: RequestSource) -> "Optional[int | _SkipSource]":
+        failure_rate = cast(FailureRate._RankingContext, ctx).failure_rate_for_rws.get(source.rse.id)
```
access the data from `cast(ctx.strategy, FailureRate).source_stats.get(source.rse.id)` directly. No need to just copy data from one dict to the other in `for_request`. If some additional computation was needed, it would be justified to implement that custom logic in `for_request`, but in this case I don't see need for complicating the program flow. You could the completely get rid of the `FailureRate._RankingContext` sub-class. And remove the override implementation of `for_request`

rcarpa commented on lib/rucio/core/transfer.py:
```diff
@@ -1156,6 +1222,7 @@ def build_transfer_paths(
         PreferDiskOverTape.external_name,
         PathDistance.external_name,
         PreferSingleHop.external_name,
+        FailureRate.external_name,
```
Don't add this strategy to the default ones. We must not change the default behavior. Users would be surprised to see the behavior of the program change so radically after an upgrade.

rcarpa commented on lib/rucio/core/transfer.py:
```diff
@@ -1071,6 +1071,69 @@ def apply(self, ctx: RequestRankingContext, source: RequestSource) -> "Optional[
         return int(len(path) > 1)
 
 
+class FailureRate(SourceRankingStrategy):
```
Add a docstring with a high level description/documentation of this strategy.

rcarpa commented:
Also, add the type hints. And fix the pyright problems (see the results of the tests pipeline run). I recommend to convert the cost to int (for example, transform it from ratio to percentage by multiplying by 100 and taking the integer part of it)

rcarpa commented on lib/rucio/core/transfer.py:
```diff
@@ -1071,6 +1071,44 @@ def apply(self, ctx: RequestRankingContext, source: RequestSource) -> "Optional[
+            self.source_stats.setdefault(stat['src_rse_id'], self._FailureRateStat()).incorporate_stat(stat)
+
+    def apply(self, ctx: RequestRankingContext, source: RequestSource) -> "Optional[int | _SkipSource]":
+        failure_rate = cast(FailureRate, ctx.strategy).source_stats.get(source.rse.id).get_failure_rate()
```
get() will return None if the element doesn't exist. This has to be handled gracefully . 
It's even shown as a mistake by the pipeline: https://github.com/rucio/rucio/actions/runs/7096607130/job/19322721914

rcarpa commented on lib/rucio/core/transfer.py:
```diff
@@ -1071,6 +1071,44 @@ def apply(self, ctx: RequestRankingContext, source: RequestSource) -> "Optional[
+    calculated by dividing files failed by files attempted.
+    """
+    class _FailureRateStat:
+        def __init__(self):
```
You are still missing some type annotations (the return values of the two __init__ . The pipeline is too strinct to my liking on this subject, but we have to live with it :)  https://github.com/rucio/rucio/actions/runs/7096607130/job/19322721542

JChenster committed:
Add FailureRate SourceRankingStrategy

We add a SourceRankingStrategy that makes source selection dynamic with respect to failure rate. We look up files done/failed in the past hour in the TransferStats schema. We assign RSEs with lower failure rates higher priority
